### DISCUSSION TITLE
How can I access packaged files from within my source generator?

### DISCUSSION THREAD

**Discussion by user72356** [dotnet/roslyn#56001] [Q&A] [answered]
I wrote a source generator that read data from a structured folder hierarchy containing about 10k files in all. I can access the files from my local system if I hardcode the location in my generator project, but I would like to ship the files alongside the generator in a NuGet package, as accompanying assets. I'm able to add them into the NuGet under "analyzers/dotnet/cs".

Now I'm stuck trying to access the assets in analyzers/dotnet/cs. It seems like they aren't deployed alongside the generator when it's running.

The only other way I can think of is adding the files as embedded resources in the generator DLL. Is that considered best practice in my use case?

Thanks!

**Comment by PathogenDavid** [comment, marked as answer]
The easiest way to accomplish this would probably be to expose the path to your assets through MSBuild. You can read the full details in [The Source Generator Cookbook](https://github.com/dotnet/roslyn/blob/d9a6dd53f7769cf80a0191b3a61b4609e2d385f8/docs/features/source-generators.cookbook.md#consume-msbuild-properties-and-metadata). (See also [the documentation on adding MSBuild props/targets to NuGet packages](https://docs.microsoft.com/en-us/nuget/create-packages/creating-a-package#include-msbuild-props-and-targets-in-a-package).)

---------

Basically you create a MSBuild props file in your NuGet package placed in the `build` directory along these lines:

```xml
<Project>
  <PropertyGroup>
    <User72356GeneratorAssets>$(MSBuildThisFileDirectory)../generator-assets/</User72356GeneratorAssets>
  </PropertyGroup>
  <ItemGroup>
    <CompilerVisibleProperty Include="User72356GeneratorAssets" />
  </ItemGroup>
</Project>
```

(The above assumes you'll put your assets in their own folder called `generator-assets`. You could keep using `analyzer/dotnet/cs` too, it's just my personal preference to separate things.)

Then from your generator you can access the path using:

```csharp
context.AnalyzerConfigOptions.GlobalOptions.TryGetValue("build_property.User72356GeneratorAssets", out string? assetsPath);
```

**Reply by user72356** [reply]
Thank you, that worked really well!

Related question, since I have no experience with msbuild other than the bare minimum... How can I access build_property.User72356GeneratorAssets from another project in my generator solution? Basically I have a test project which has a project reference on the generator, in order to test the generator as I go. The property isn't accessible to the test project obviously.

**Reply by PathogenDavid** [reply]
> How can I access build_property.User72356GeneratorAssets from another project in my generator solution?

To answer this question literally: Getting properties or items from one project to another is not for the faint of heart. Projects are effectively built in isolation so getting properties/items between them isn't super efficient\* or fun. (It basically involves executing a special target in the remote project using the [`MSBuild` task](https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild-task?view=vs-2019) and getting it back that way. I don't have a concise example to share, I could try to throw something together if you're curious though.)

(\*I assume, I've never measured.)

-----------------

As far as getting the actual path in your tests goes, the easiest way would be either to
1. ~~Copy the generator assets to the tests' output directory via `CopyToOutputDirectory="PreserveNewest"` and build a path based on the working directory.~~ (This would normally be my personal preference, but in your case there's a ton of files so this isn't super ideal.)
2. Figure out the path to the assets folder relative to the test's output directory based on either `AppContext.BaseDirectory` or `Environment.CurrentWorkingDirectory`. (The main issue here is the number of nested folders in your `bin` directory changes based on your build settings. Regardless this is probably still my preference in your situation, just make it tolerant to an arbitrary level of nesting.)

-----------------

If 1 isn't ideal because the assets are large and/or 2 feels too brittle to you, you can use an assembly attribute to get the property from MSBuild into your test.

The below assumes a solution layout along these lines:

* `MySourceGenerator`
  * `MySourceGenerator.csproj`
  * `generator-assets`
    * `SomeGeneratorAsset1.json`
    * ...
    * `SomeGeneratorAsset10000.json`
* `MyTestAssembly`
  * `MyTestAssembly.csproj`
  * `MyTestFacts.cs`

In `MyTestAssembly` define an attribute that will hold your asset path:

```csharp
using System;

namespace MyTestAssembly
{
    [AttributeUsage(AttributeTargets.Assembly)]
    internal sealed class User72356GeneratorAssetsPathAttribute : Attribute
    {
        public string Path { get; }

        public User72356GeneratorAssetsPathAttribute(string path)
            => Path = path;
    }
}
```

In `MyTestAssembly.csproj` add your property and an `AssemblyAttribute` item:

```xml
  <PropertyGroup>
    <User72356GeneratorAssets>$(MSBuildThisFileDirectory)../MySourceGenerator/generator-assets/</User72356GeneratorAssets>
  </PropertyGroup>

  <ItemGroup>
    <AssemblyAttribute Include="MyTestAssembly.User72356GeneratorAssetsPathAttribute">
      <_Parameter1>$(User72356GeneratorAssets)</_Parameter1>
    </AssemblyAttribute>
  </ItemGroup>
```

With all that in place, you can access the value of `User72356GeneratorAssets` (in the context of the test assembly) using:

```csharp
typeof(MyTestFacts).Assembly.GetCustomAttribute<User72356GeneratorAssetsPathAttribute>()!.Path
```

(Note: I'm pretty sure that this will have a quirk of breaking your tests if you move the solution elsewhere on the filesystem without rebuilding since the up-to-date checks won't notice the value of the property has changed. Also this assumes you won't publish your test assemblies since it results in a system-specific path being embedded in them. I've never heard of anyone doing that, but thought I should point it out.)

**Reply by user72356** [reply]
Thank you VERY MUCH for taking the time! This was really helpful. I'm going to work on this right away, have a good one!

**Reply by PathogenDavid** [reply]
No problem, you too!

Also I almost forgot to mention: If you find yourself exploring more MSBuild stuff, [the MSBuild Structured Log Viewer](https://github.com/KirillOsenkov/MSBuildStructuredLog) and [Project System Tools Extension](https://github.com/dotnet/project-system-tools) are both invaluable tools. (The extension is specifically useful for capturing build logs from Visual Studio. It's particularly useful for debugging stuff specific to design-time builds.)
